Merge various fixes for C-LCOW since conf-aci/0.2.5 (part 3)

micromaomao opened the pull request:
Part 1: https://github.com/microsoft/hcsshim/pull/2581 (merged) \
Part 2: https://github.com/microsoft/hcsshim/pull/2768 (merged) \
This is part 3.

This PR merges a set of changes, all already reviewed by ContainerPlatform team except 2, applied to the C-ACI release.

- [**Merged PR 12878148: bridge: Force sequential message handling for confidential containers**](https://github.com/microsoft/hcsshim/pull/2559/commits/894e19b1c9027d7d8217f829e73057cf6bc0641c) ([original ADO PR](https://dev.azure.com/msazure/ContainerPlatform/_git/Microsoft.hcsshim/pullrequest/12878148))
- [**Merged PR 13618357: guest/network: Restrict hostname to valid characters**](https://github.com/microsoft/hcsshim/pull/2559/commits/f199f08f8666d93bd42f0898503dddefb9699f4a) ([original ADO PR](https://dev.azure.com/msazure/ContainerPlatform/_git/Microsoft.hcsshim/pullrequest/13618357))
- [**Merged PR 12878779: Enhance uvm_state::hostMounts to track in-use mounts, and prevent unmounting or deleting in-use things**](https://github.com/microsoft/hcsshim/pull/2559/commits/aec527741c6fd64c4b47a9681d383ce67002700e) ([original ADO PR](https://dev.azure.com/msazure/ContainerPlatform/_git/Microsoft.hcsshim/pullrequest/12878779))
  - [**gcs: tighten hcsv2 container lifecycle and mount checks**](https://github.com/microsoft/hcsshim/pull/2559/commits/e22b4dec63256bcf5d935f14415e69e18966cb3e) (new)
- [**Merged PR 13627088: guest: Don't allow host to set mount options**](https://github.com/microsoft/hcsshim/pull/2559/commits/d429219ceb1da4e14b0b33ee8086a00e890ce5c5) ([original ADO PR](https://dev.azure.com/msazure/ContainerPlatform/_git/Microsoft.hcsshim/pullrequest/13627088))
- [**Merged PR 13694574: hcsv2/uvm, rego: Enforce OCI spec does not contain any LinuxDevices**](https://github.com/microsoft/hcsshim/pull/2559/commits/36c21c1e74925900defa5e67194fe6d70f03084f) ([original ADO PR](https://dev.azure.com/msazure/ContainerPlatform/_git/Microsoft.hcsshim/pullrequest/13694574))

micromaomao committed:
Merged PR 13618357: guest/network: Restrict hostname to valid characters

[cherry-picked from 055ee5eb4a802cb407575fb6cc1e9b07069d3319]

guest/network: Restrict hostname to valid characters

Because we write this hostname to /etc/hosts, without proper validation the host
can trick us into writing arbitrary data to /etc/hosts, which can, for example,
redirect things like ip6-localhost (but likely not localhost itself) to an
attacker-controlled IP address.

We implement a check here that the host-provided DNS name in the OCI spec is
valid.

ACI actually restricts this to 5-63 characters of a-zA-Z0-9 and '-', where the
first and last characters can not be '-'.  This aligns with the Kubernetes
restriction.  c.f. IsValidDnsLabel in Compute-ACI.  However, there is no
consistent official agreement on what a valid hostname can contain.  RFC 952
says that "Domain name" can be up to 24 characters of A-Z0-9 '.' and '-', RFC
1123 expands this to 255 characters, but RFC 1035 claims that domain names can
in fact contain anything if quoted (as long as the length is within 255
characters), and this is confirmed again in RFC 2181.  In practice we see names
with underscopes, most commonly \_dmarc.example.com.  curl allows 0-9a-zA-Z and
-.\_|~ and any other codepoints from \u0001-\u001f and above \u007f:
https://github.com/curl/curl/blob/master/lib/urlapi.c#L527-L545

With the above in mind, this commit allows up to 255 characters of a-zA-Z0-9 and
'_', '-' and '.'.  This change is applied to all LCOW use cases.

This fix can be tested with the below code to bypass any host-side checks:

	+++ b/internal/hcsoci/hcsdoc_lcow.go
	@@ -52,6 +52,10 @@ func createLCOWSpec(ctx context.Context, coi *createOptionsInternal) (*specs.Spe
			spec.Linux.Seccomp = nil
		}

	+	if spec.Annotations[annotations.KubernetesContainerType] == "sandbox" {
	+		spec.Hostname = "invalid-hostname\n1.1.1.1 ip6-localhost ip6-loopback localhost"
	+	}
	+
		return spec, nil
	}

Output:

	time="2025-10-01T15:13:41Z" level=fatal msg="run pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: failed to create container f2209bb2960d5162fc9937d3362e1e2cf1724c56d1296ba2551ce510cb2bcd43: guest RPC failure: hostname \"invalid-hostname\\n1.1.1.1 ip6-localhost ip6-loopback localhost\" invalid: must match ^[a-zA-Z0-9_\\-\\.]{0,999}$: unknown"

Related work items: #34370598
Closes: https://msazure.visualstudio.com/One/_workitems/edit/34370598
Signed-off-by: Tingmao Wang <tingmaowang@microsoft.com>

micromaomao committed:
Merged PR 12878148: bridge: Force sequential message handling for confidential containers

[cherry-picked from f81b450894206a79fff4d63182ff034ba503ebdb]

This PR contains 2 commits. The first one is the fix:

**bridge: Force sequential message handling for confidential containers**

This fixes a vulnerability (and reduces the surface for other similar potential
vulnerabilities) in confidential containers where if the host sends a
mount/unmount modification request concurrently with an ongoing CreateContainer
request, the host could re-order or skip image layers for the container to be
started.

While this could be fixed by adding mutex lock/unlock around the individual
modifyMappedVirtualDisk/modifyCombinedLayers/CreateContainer functions, we
decided that in order to prevent any more of this class of issues, the UVM, when
running in confidential mode, should just not allow concurrent requests (with
exception for any actually long-running requests, which for now is just
waitProcess).

The second one adds a log entry for when the processing thread blocks.  This
will make it easier to debug should the gcs "hung" on a request.

This PR is created on ADO targeting the conf branch as this security
vulnerability is not public yet. This fix should be backported to main once
deployed.

Fix usage of IsSNP() to handle the new error return value

Related work items: #33357501, #34327300
Signed-off-by: Tingmao Wang <tingmaowang@microsoft.com>

micromaomao committed:
Merged PR 12878779: Enhance uvm_state::hostMounts to track in-use mounts, and prevent unmounting or deleting in-use things

[cherry-picked from d0334883cd43eecbb401a6ded3e0317179a3e54b]

This set of changes adds some checks (when running with a confidential policy)
to prevent the host from trying to clean up mounts, overlays, or the container
states dir when the container is running (or when the overlay has not been
unmounted yet).  This is through enhancing the existing `hostMounts` utility, as
well as adding a `terminated` flag to the Container struct.

The correct order of operations should always be:

- mount read-only layers and scratch (in any order, and individual containers
(not the sandbox) might not have their own scratch) - mount the overlay - start
the container - container terminates - unmount overlay - unmount read-only
layers and scratch

The starting up order is implied, and we now explicitly deny e.g. unmounting
layer/scratch before unmounting overlay, or unmounting the overlay while
container has not terminated.

We also deny deleteContainerState requests when the container is running or when
the overlay is mounted.  Doing so when a container is running can result in
unexpectedly deleting its files, which breaks it in unpredictable ways and is
bad.

Signed-off-by: Tingmao Wang <tingmaowang@microsoft.com>

micromaomao committed:
Merged PR 13627088: guest: Don't allow host to set mount options

[cherry-picked from 1dd0b7ea0b0f91d3698f6008fb0bd5b0de777da6]

Blocks mount option passing for 9p (which is accidental) and SCSI disks.

- guest: Restrict plan9 share names to digits only on Confidential mode
- hcsv2/uvm: Restrict SCSI mount options in confidential mode
   (The only one we allow is `ro`)

Related work items: #34370380
Signed-off-by: Tingmao Wang <tingmaowang@microsoft.com>

micromaomao committed:
Merged PR 13694574: hcsv2/uvm, rego: Enforce OCI spec does not contain any LinuxDevices

[cherry-picked from 9f69e49f3..72b338a99]
[forward-ported from original PR 13653011 (e631aa42c6039ab0d228b746b280c26809433f00)]

See https://msazure.visualstudio.com/ContainerPlatform/_git/Microsoft.hcsshim/pullrequest/13653011

Signed-off-by: Tingmao Wang <tingmaowang@microsoft.com>

anmaxvl commented on internal/guest/runtime/hcsv2/uvm.go:
```diff
@@ -109,6 +110,16 @@ type Host struct {
 	// hostMounts keeps the state of currently mounted devices and file systems,
 	// which is used for GCS hardening.
 	hostMounts *hostMounts
+	// A permanent flag to indicate that further mounts, unmounts and container
```
comment format nit (start with the name):
```
// mountsBroken is a permanent flag ...
```

anmaxvl replied:
can't recall if we've ever ran into broken mounts for non-confidential LCOW. I think most of the time it's the block device that's busy and we fail to unmount or remove the verity target and this breaks. Either way, consider the following:
```go
type uvmConsistencyError struct {
    mu    sync.Mutex
    cause string // empty = healthy
}

func (u *uvmConsistencyError) Set(cause string) {
    u.mu.Lock()
    defer u.mu.Unlock()
    if u.cause == "" {
        u.cause = cause
    }
}

func (u *uvmConsistencyError) Check() error {
    u.mu.Lock()
    defer u.mu.Unlock()
    if u.cause != "" {
        return fmt.Errorf("operation blocked: UVM in inconsistent state (%s)", u.cause)
    }
    return nil
}
```
On `Host`
```go
type Host struct {
    // ...
    uvmError uvmConsistencyError
}
```

On call sites:
```go
// Record failure
err = scsi.Unmount(...)
if err != nil {
    h.uvmError.Set(fmt.Sprintf("unmount scsi at %s: %v", mvd.MountPath, err))
    return err
}

// Gate operation — only matters in confidential mode
func (h *Host) checkState() error {
    if h.HasSecurityPolicy() {
        return h.uvmError.Check()
    }
    return nil
}
```

micromaomao replied:
Nit addressed in 614e5dc16

micromaomao replied:
Refactor suggestion implemented in dadba5ead

micromaomao commented:
@anmaxvl Have you had a chance to review? This will end up blocking a C-WCOW release as it will need to also release C-LCOW from main in the same package to avoid having separate clusters

anmaxvl commented on internal/guest/runtime/hcsv2/uvm.go:
```diff
@@ -386,7 +398,44 @@ func checkContainerSettings(sandboxID, containerID string, settings *prot.VMHost
+	}).Error("Host::mountsBroken set to true. All further mounts/unmounts, container creation and deletion will fail.")
+}
+
+func checkExists(path string) (error, bool) {
```
go convention is to return error last, i.e. `func checkExists(path string) (bool, error)`

micromaomao replied:
Addressed in 614e5dc16

anmaxvl commented on internal/guest/runtime/hcsv2/uvm.go:
```diff
@@ -386,7 +398,44 @@ func checkContainerSettings(sandboxID, containerID string, settings *prot.VMHost
+		return
+	}
+	h.mountsBroken.Store(true)
+	h.mountsBrokenCausedBy = cause
```
this could technically race if two go routines are trying to set/access this.

anmaxvl replied:
```go
func (h *Host) setMountsBrokenIfConfidential(cause string) {
    if !h.HasSecurityPolicy() {
        return
    }
    h.mountsBrokenOnce.Do(func() {
        h.mountsBrokenCausedBy = cause  // write before atomic store
        h.mountsBroken.Store(true)
        log.G(context.Background()).WithField("cause", cause).Error(
            "Host::mountsBroken set to true. ...")
    })
}
```

micromaomao replied:
Addressed in 614e5dc16

anmaxvl commented on internal/gcs/unrecoverable_error.go:
```diff
@@ -0,0 +1,49 @@
+		"Unrecoverable error in GCS: %v\n%s",
+		err, stackTrace,
+	)
+	isSnp := amdsevsnp.IsSNP()
```
`IsSNP()` returns `(bool, error)`, which isn't handled here: https://github.com/microsoft/hcsshim/blob/main/pkg/amdsevsnp/report_linux.go#L95

micromaomao replied:
It only errors on Windows, for the IsSNP() in report_linux.go it is not possible for it to return with `err == nil`. What is your suggestion as to how to handle errors? (Treat as non-SNP, as opposed to what's currently done in this PR?)

micromaomao replied:
wait, have you seen https://github.com/microsoft/hcsshim/pull/2559/commits/91023202d0670eaa38560536da829421873d1c68?

anmaxvl commented on pkg/securitypolicy/securitypolicyenforcer_rego.go:
```diff
@@ -1191,3 +1197,81 @@ func (policy *regoEnforcer) EnforceRegistryChangesPolicy(ctx context.Context, co
 	return GetAllUserInfo(process, rootPath)
 }
+
+type revertableSectionHandle struct {
```
How about we add a transaction interface to the enforcer:
```go
type SecurityPolicyEnforcer interface {
    // ...existing methods...

    // WithTransaction snapshots policy metadata, runs fn, and rolls back
    // on error. Nested/concurrent calls return an error.
    WithTransaction(fn func() error) error
}
```
Then add corresponding implementations:
```go
// Open door — passthrough
func (*OpenDoorSecurityPolicyEnforcer) WithTransaction(fn func() error) error {
    return fn()
}

// Closed door — passthrough
func (*ClosedDoorSecurityPolicyEnforcer) WithTransaction(fn func() error) error {
    return fn()
}

// Rego — snapshot/restore
func (policy *regoEnforcer) WithTransaction(fn func() error) error {
    if !policy.transactionLock.TryLock() {
        return fmt.Errorf("nested or concurrent policy transactions are not supported")
    }
    defer policy.transactionLock.Unlock()

    saved, err := policy.rego.SaveMetadata()
    if err != nil {
        return fmt.Errorf("failed to snapshot policy metadata: %w", err)
    }

    err = fn()

    if err != nil {
        if restoreErr := policy.rego.RestoreMetadata(saved); restoreErr != nil {
            gcs.UnrecoverableError(fmt.Errorf(
                "failed to rollback policy metadata after error (%w): %v",
                err, restoreErr,
            ))
        }
        log.G(context.Background()).WithError(err).Warn(
            "rolled back policy metadata due to error")
    }

    return err
}
```

The call site becomes:
```go
func (h *Host) modifyMappedVirtualDisk(
    ctx context.Context,
    rt guestrequest.RequestType,
    mvd *guestresource.LCOWMappedVirtualDisk,
) error {
    // ...setup verityInfo...

    return h.securityOptions.PolicyEnforcer.WithTransaction(func() error {
        switch rt {
        case guestrequest.RequestTypeAdd:
            if err := securityPolicy.EnforceDeviceMountPolicy(ctx, mvd.MountPath, deviceHash); err != nil {
                return err
            }
            return scsi.Mount(...)
        case guestrequest.RequestTypeRemove:
            if err := securityPolicy.EnforceDeviceUnmountPolicy(ctx, mvd.MountPath); err != nil {
                return err
            }
            if err := scsi.Unmount(...); err != nil {
                h.uvmError.Set(fmt.Sprintf("unmount scsi at %s: %v", mvd.MountPath, err))
                return err
            }
            return nil
        }
    })
}
```

micromaomao replied:
I'm happy to implement this. Did you mean to change `policy.revertableSectionLock.Lock()` to `policy.transactionLock.TryLock()`? (i.e. the caller now has to make sure not to call this concurrently, which given the bridge sequential processing, should indeed not happen.)

micromaomao replied:
Implemented in 3552fdbde

anmaxvl reviewed (changes requested):
Left comments on metadata commit. Lets' have a separate PR to address them.

anmaxvl commented on internal/guest/bridge/bridge.go:
```diff
@@ -177,6 +177,10 @@ type Bridge struct {
+	// Setting ForceSequential to true will force the bridge to only process one
+	// request at a time, except for certain long-running operations (as defined
+	// in asyncMessages).
+	ForceSequential bool
```
nit: name it `Sequential`

micromaomao replied:
Fixed in e48da68ca

anmaxvl commented on cmd/gcs/main.go:
```diff
@@ -414,6 +415,16 @@ func main() {
+	forceSequential, err := amdsevsnp.IsSNP()
+	if err != nil {
+		// IsSNP cannot fail on LCOW
+		logrus.Errorf("Got unexpected error from IsSNP(): %v", err)
```
```
log.G(context.TODO()).WithError(err).Error("unexpected error from IsSNP")
```

micromaomao replied:
Fixed in e48da68ca

anmaxvl commented on internal/guest/bridge/bridge.go:
```diff
@@ -256,6 +260,14 @@ func (b *Bridge) ShutdownRequested() bool {
+// that in sequential mode, these messages will still wait for any in-progress
+// non-async messages to be handled before they are processed, but once they are
+// "acknowledged", the rest will be done asynchronously.
+var alwaysAsyncMessages map[prot.MessageIdentifier]bool = map[prot.MessageIdentifier]bool{
```
with only a single entry, consider:
```go
func alwaysAsync(msgType prot.MessageIdentifier) bool {
    return msgType == prot.ComputeSystemWaitForProcessV1
}
```

micromaomao replied:
why is a function better?

micromaomao replied:
Fixed with rebase

anmaxvl commented on internal/guest/bridge/bridge.go:
```diff
@@ -409,30 +425,36 @@ func (b *Bridge) ListenAndServe(bridgeIn io.ReadCloser, bridgeOut io.WriteCloser
-						oc.SetSpanStatus(span, err)
-					}
-					setErrorForResponseBase(resp.Base(), err, "gcs" /* moduleName */)
+		doOneRequest := func(r *Request) {
```
nit: `doRequest`

micromaomao replied:
Fixed in e48da68ca

anmaxvl commented on internal/guest/bridge/bridge.go:
```diff
@@ -409,30 +425,36 @@ func (b *Bridge) ListenAndServe(bridgeIn io.ReadCloser, bridgeOut io.WriteCloser
+
+		for req := range requestChan {
+			if b.ForceSequential && !alwaysAsyncMessages[req.Header.Type] {
+				runSequentialRequestHandler(req, doOneRequest)
```
nit: `runSequentialRequest` and leave a comment that it'll log warn after 5 seconds.

micromaomao replied:
Fixed in e48da68ca

anmaxvl commented on internal/guest/runtime/hcsv2/process.go:
```diff
@@ -99,6 +99,7 @@ func newProcess(c *Container, spec *oci.Process, process runtime.Process, pid ui
 			log.G(ctx).WithError(err).Error("failed to wait for runc process")
 		}
 		p.exitCode = exitCode
+		c.terminated.Store(true)
```
you need to make sure that this is actually an init process before you set this to `true`.

micromaomao replied:
Thanks a lot for spotting! Fixed with rebase.

anmaxvl commented on internal/guest/runtime/hcsv2/uvm.go:
```diff
@@ -1319,11 +1337,42 @@ func (h *Host) modifyMappedVirtualDisk(
 					if err != nil {
 						return errors.Wrapf(err, "mounting scsi device controller %d lun %d onto %s denied by policy", mvd.Controller, mvd.Lun, mvd.MountPath)
 					}
+					if h.hostMounts != nil {
```
consider adding an interface instead of `if h.hostMounts != nil` everywhere.
```go
type hostMountsTracker interface {
    Lock()
    Unlock()
    AddRODevice(mountPath, sourcePath string) error
    RemoveRODevice(mountPath, sourcePath string) error
    AddRWDevice(mountPath, sourcePath string, encrypted bool) error
    RemoveRWDevice(mountPath, sourcePath string, encrypted bool) error
    AddOverlay(mountPath string, layers []string, scratchDir string) error
    RemoveOverlay(mountPath string) (func(), error)
    IsEncrypted(path string) bool
    HasOverlayMountedAt(path string) bool
}

type noOpHostMounts struct{}
func (n *noOpHostMounts) Lock()    {}
func (n *noOpHostMounts) Unlock()  {}
func (n *noOpHostMounts) AddRODevice(...) error { return nil }
// ... rest as no-ops

type realHostMounts struct { ... } // current implementation
```
This way we can default to a no-op and when securitypolicy is present in SetSecurityOptions, we assign a real implementation.

micromaomao replied:
That would result in the following 15 lines doing nothing in non-confidential only by implication, and makes it harder to follow code as "find definition" now doesn't find you the actual definition automatically.

Do you still want to change it this way?

anmaxvl committed:
gcs: tighten hcsv2 container lifecycle and mount checks

Move container lifecycle tracking to explicit states
(created, running, terminated) and use those states
for container access, overlay usage, and confidential
delete-state guards.

Fix overlay mount path normalization so add/remove
works reliably with clean vs unclean paths.

Rename GetCreatedContainer to GetInitializedContainer
and update callers/tests.

Add lifecycle and path-normalization regression tests.

Assisted-by: Github Copilot
Signed-off-by: Maksim An <maksiman@microsoft.com>

micromaomao commented on internal/guest/runtime/hcsv2/uvm_state.go:
```diff
@@ -4,91 +4,362 @@
+func (hm *hostMounts) AddOverlay(mountPath string, layers []string, scratchDir string) (err error) {
+	hm.expectLocked()
+
+	cleanMountPath := filepath.Clean(mountPath)
```
I'm not sure if this is a good idea but will leave it for now (it shouldn't matter for confidential)
